### PR TITLE
[MRG] use selectors to get output and data from mpi_child.py

### LINKED ISSUE #192: Improve passing of simulation results from mpi_child.py to MPIBackend
Since `mpi_child.py` is a separate process, the ways to send data back and forth are limited. Right now, data is being packed into stderr, but complications can arise when MPI outputs an error in the middle of data generation that also goes to stderr.

Fix this with a defined protocol for sending both data and error messages between processes. Using file descriptor pipes may not be the best way.

Found from PR #190

### PR DESCRIPTION
**Pull request by blakecaldwell**
This uses the well-known `select()` mechanism to wait for input from `mpi_child.py` in MPIBackend. A signal is sent by the child (string "sim_complete") to let MPIBackend know that it has received all output that should be printed to the screen (stdout and stderr) and to expect simulation data as a pickled object.

Resolves #192

### CODE REVIEW

**Comment by blakecaldwell** [comment]
Note that there is nothing new to test here. It actually reduces the number of special cases and the existing test is more general (same across different platforms).

File: hnn_core/mpi_child.py [hunk 1]
```diff
@@ -23,18 +23,6 @@ def run_mpi_simulation():
...
-
-    # suppress output to stderr
-    stderr_fileno = sys.stderr.fileno()
-    null_fd = os.open(os.devnull, os.O_RDWR)
-    old_err_fd = os.dup(stderr_fileno)
-    os.dup2(null_fd, stderr_fileno)
-
-    # temporarily use a StringIO object to capture stderr
-    str_err = io.StringIO()
-    sys.stderr = str_err
```
Comment (blakecaldwell) [inline hnn_core/mpi_child.py:37]: All of this output stream redirection is gone. Yay!

File: hnn_core/mpi_child.py [hunk 2]
```diff
@@ -76,38 +64,22 @@ def run_mpi_simulation():
...
     if rank == 0:
-        data_str = data_iostream.getvalue().decode()
-        sys.stderr.write(data_str)
-
-    # close the StringIO object
-    str_err.close()
+        # the parent process is waiting for the string "sim_complete"
+        # to signal that the output will only contain sim_data
+        sys.stdout.write('sim_complete')
+        sys.stdout.flush()  # flush to ensure signal is not buffered
```
Comment (blakecaldwell) [inline hnn_core/mpi_child.py:75]: Here's the signal being sent to MPIBackend. The `flush()` is important.

File: hnn_core/parallel_backends.py [hunk 3]
```diff
@@ -11,7 +11,7 @@
 import codecs
 from warnings import warn
 from subprocess import Popen
-import select
+import selectors
```
Comment (blakecaldwell) [inline hnn_core/parallel_backends.py:14]: [`selectors`](https://docs.python.org/3/library/selectors.html) is a higher-level module than `select`

File: hnn_core/parallel_backends.py [hunk 4]
```diff
@@ -223,6 +223,28 @@ def __exit__(self, type, value, traceback):
...
+        # data is signalled by the process terminating
+        return False
+
+    def _read_stdout(self, fd, mask):
+        """read from fd until receiving the process simulation is complete"""
+        data = os.read(fd, 4096)
+        if data:
+            str_data = data.decode()
+            if str_data == 'sim_complete':
+                return True
```
Comment (blakecaldwell) [inline hnn_core/parallel_backends.py:241]: Here's where the signal is recognized.

File: hnn_core/parallel_backends.py [hunk 5]
```diff
@@ -282,67 +304,51 @@ def simulate(self, net):
...
+            for key, mask in events:
+                callback = key.data
+                finished = callback(key.fileobj, mask)
+                if finished:
+                    # finishied receiving printable output
+                    self.sel.unregister(pipe_stdout_r)
+                    self.sel.unregister(pipe_stderr_r)
+                    # everything received should now be handled by _read_data
+                    self.sel.register(pipe_stderr_r, selectors.EVENT_READ,
+                                      self._read_data)
```
Comment (blakecaldwell) [inline hnn_core/parallel_backends.py:332]: Before this, `_read_stdout` was called for events on stderr and stdout. Now just `_read_data` is called to handle events on stderr

File: hnn_core/parallel_backends.py [hunk 6]
```diff
@@ -282,67 +304,51 @@ def simulate(self, net):
...
-        else:
-            sim_failed = True
-            if len(proc_stderr) > 0:
-                mpi_error_msg = proc_stderr
-
-        if sim_failed or len(mpi_error_msg) > 0:
-            sys.stderr.write("MPI simulation returned extra messages:\n")
-            sys.stderr.write(mpi_error_msg)
-            sys.stderr.write("\n*** End MPI messages ***\n\n")
-
```
Comment (blakecaldwell) [inline hnn_core/parallel_backends.py:336]: This was the code to search for padding of the data and separate out error messages/exceptions. Yuck.

File: hnn_core/tests/test_compare_hnn.py [hunk 7]
```diff
@@ -124,15 +123,10 @@ def test_mpi_failure():
     # this MPI paramter will cause a MPI job with more than one process to fail
     environ["OMPI_MCA_btl"] = "self"
 
-    # this is a bummer, but pytest on linux does something w/ file descriptors
-    if 'linux' in sys.platform:
```
Comment (blakecaldwell) [inline hnn_core/tests/test_compare_hnn.py:128]: Since all output goes to stdout now, we can just use `redirect_stdout`, which works correctly in Linux.

**Review by blakecaldwell — COMMENTED** [review]
This is ready to go by me. Added review comments to point out key code sections.

**Comment by jasmainak** [comment]
Thanks @blakecaldwell !

**Pull request merged**